### PR TITLE
[MRG] add `sourmash sig grep`

### LINKED ISSUE #1075: add a `sig grep` command?
It'd be nice to be able to find signatures in files by md5... and also search by name, filename, etc.

ref #993

### PR DESCRIPTION
**Pull request by ctb**
Implements `sourmash sig grep` as in https://github.com/sourmash-bio/sourmash/issues/1075.

`sig grep` does substring matching on name, filename, and md5. It supports regular expressions, as well as the `-v` (exclude instead of include) and `-i` (case insensitive) options. It also supports CSV output in manifest format (for use as picklists), and the `-c/--count` option to just count the number of signatures in each file.

A key feature of `grep` is that it (by default) _requires_ manifests. That way users won't get caught up accidentally in searches of very large databases w/o manifests, which has been happening to me a lot recently.

Fixes #1075.

## Examples

Extract by acccession:
```
% sourmash sig grep GCF_000019185. /group/ctbrowngrp/gtdb/databases/ctb/gtdb-rs202.genomic.k31.zip -o out.zip

...

loaded 258406 total that matched ksize & molecule type
extracted 1 signatures from 1 file(s)
```

Fail when no manifest present:
```
% sourmash sig grep GCF_000019185. /group/ctbrowngrp/gtdb/databases/gtdb-rs202.genomic.k31.zip -o out.zip

...

ERROR on filename '/group/ctbrowngrp/gtdb/databases/gtdb-rs202.genomic.k31.zip'.
sig grep requires a manifest by default, but no manifest present.
```

Count some things and output a manifest:
```
% sourmash sig grep -ci shewanella /group/ctbrowngrp/gtdb/databases/ctb/gtdb-rs202.genomic.k31.zip --csv matches.csv

== This is sourmash version 4.2.4.dev27+g0a1b713f. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

(no signatures will be output because of --silent/--count).
252 matches: /group/ctbrowngrp/gtdb/databases/ctb/gtdb-rs202.genomic.k31.zip
wrote manifest containing matches to CSV matches 'matches.csv'
```

## Notes

eventually we should probably add the default manifest requirement to several more sig subcommands, like extract, but this will require a major version bump.

## TODO

- [x] upgrade PR description with example command
- [x] tests needed!
- [x] Visit #1075 and extract non-grep stuff to new issue
- [ ] make issue to require manifests in various sig subcomands (sig extract, for example)
- [x] write docs

### CODE REVIEW

**Comment by ctb** [comment]
@taylorreiter @bluegenes @luizirber any thoughts on additional functionality beyond regexps, -v, and -i? this is already an immediately useful command 😆

File: src/sourmash/sig/grep.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            total_rows_examined += 1
+
+        # convert to picklist, grabme.
+        sub_manifest = CollectionManifest(sub_rows)
+        sub_picklist = sub_manifest.to_picklist()
+
+        try:
+            idx = idx.select(picklist=sub_picklist)
+        except ValueError:
+            error("** This input collection doesn't support 'extract' with picklists.")
```
Comment (bluegenes) [inline src/sourmash/sig/grep.py]: `extract` --> `grep`?

**Comment by bluegenes** [comment]
> @taylorreiter @bluegenes @luizirber any thoughts on additional functionality beyond regexps, -v, and -i? this is already an immediately useful command 😆

Maybe also `-c`/`--count` for just counting results you would get (suppressing normal matched output)?

**Comment by ctb** [comment]
> Maybe also `-c`/`--count` for just counting results you would get (suppressing normal matched output)?

yeah, I like this! 

what about other outputs, like `-l` for showing what files things are in? I can see it maybe being useful.

and another brainstorming thought - I'm guessing it might be handy to support switching the output to a manifest-style CSV so you can use this to create picklists.

**Comment by ctb** [comment]
Added `-c/--count` and `--csv` for outputting manifests of matches.

I think (modulo tests and some refactoring) this PR might be feature complete, at least until we get some experience in using the functionality and have more ideas on what to add. I'll push forward on the tests as I have time, but it's pretty usable already, so let me know if you have any thoughts or comments!

**Comment by ctb** [comment]
Welp, got the tests and docs done sooner than expected. Ready for review & merge @sourmash-bio/devs!

File: src/sourmash/cli/sig/grep.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+displayed by `sig describe`.
+
+'pattern' can be a string or a regular expression.
+
+'sig grep' uses the built-in Python regexp module, 're', to implement
+regexp searching. See https://docs.python.org/3/howto/regex.html and
+https://docs.python.org/3/library/re.html for details.
+
+The '-v' (exclude) and '-i' (case-insensitive) options of 'grep' are
+supported.
```
Comment (bluegenes) [inline src/sourmash/cli/sig/grep.py]:
```suggestion
The '-v' (exclude), '-i' (case-insensitive), and `-c` (count) options of 'grep' are
supported.
```

File: src/sourmash/sig/grep.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        else:
+            error("no matching signatures found!")
+            sys.exit(-1)
+
+    if args.csv:
+        notify(f"wrote manifest containing all matches to CSV file '{args.csv}'")
+        csv_obj.close()
+
+    if picklist:
+        sourmash_args.report_picklist(args, picklist)
```
Comment (bluegenes) [inline src/sourmash/sig/grep.py:129]: does this need a test for grep + picklist? I see one for when this will fail (LCA)...

**Review by bluegenes — APPROVED** [review]
It looks like `sig grep` works with picklist input as well - picklist is applied first, then pattern match. If yes, I think we need a test for this case (pattern + picklist input). Happy to review again with add'l test, or leave to your discretion.

File: src/sourmash/sig/grep.py (hunk 3 above)
Comment (ctb) [reply src/sourmash/sig/grep.py:129]: ooh, nice catch! `test_sig_grep_7_picklist_md5_lca_fail` tests the failure, but there's no test for the successful use of `--picklist`. Will add.

Comment (ctb) [reply src/sourmash/sig/grep.py:129]: added in 76b2b02f

Comment (bluegenes) [reply src/sourmash/sig/grep.py:129]: looks good!

**Comment by ctb** [comment]
🎉

**Pull request merged**